### PR TITLE
Supports SOL-180. Allows for ids to be excluded from a search.

### PR DESCRIPTION
**Pull request by martynjames**
Also, a couple drive-by items

### CODE REVIEW

**Comment by martynjames** [comment]
@dino-cikatic - Please take a review pass on this

**Comment by martynjames** [comment]
@dino-cikatic; @e-kolpakov - This has the stuff to support SOL-174 in the edx-search app. There are some merge conflicts that are resolved in `mjames/test_merge` branch here. I'll update this PR with the merged code once tests have completed.

File: search/tests/tests.py [hunk 1]
```diff
@@ -551,12 +552,70 @@ def test_pagination(self):
...
     """ For testing pieces of the Mock Engine that have no equivalent in Elastic """
 
+    _searcher = None
+
+    @property
+    def searcher(self):
+        """ cached instance of search engine """
+        if self._searcher is None:
+            self._searcher = SearchEngine.get_search_engine(TEST_INDEX_NAME)
+        return self._searcher
```
Comment (e-kolpakov) [inline search/tests/tests.py]: @martynjames there's a `SearcherMixin` in test utils that do exactly that. Also, I believe pylint should complain about code duplication here and there. Actually the reason I made `SearcherMixin` is that pylint yelled at me.

File: search/tests/mock_search_engine.py [hunk 2]
```diff
@@ -64,9 +65,23 @@ def value_matches(doc, field_name, field_value):
...
+        if isinstance(field_value, DateRange) or isinstance(field_value, datetime):
+            if isinstance(compare_value, basestring):
+                compare_value = json_date_to_datetime(compare_value)
+
+            field_has_tz_info = False
+            if isinstance(field_value, DateRange):
+                if field_value.lower and field_value.lower.tzinfo is not None:
+                    field_has_tz_info = True
+                if field_value.upper and field_value.upper.tzinfo is not None:
+                    field_has_tz_info = True
```
Comment (e-kolpakov) [inline search/tests/mock_search_engine.py]: @martynjames nit: looks like it's equivalent to

    field_has_tz_info = (
        (field_value.lower and field_value.lower.tzinfo is not None) or
        (field_value.upper and field_value.upper.tzinfo is not None)
    )

or

    has_tz = lambda value = value and value.tzinfo is not None
    field_has_tz_info = has_tz(field_value.lower) or has_tz(field_value.upper) 

which I find more concise. A matter of taste though.

File: search/tests/tests.py [hunk 3]
```diff
@@ -625,6 +684,46 @@ def test_datetime_conversion(self):
         json_datetime = "2015-01-31T07:30:28.65785Z"
         self.assertTrue(json_date_to_datetime(json_datetime), datetime(2015, 1, 31, 7, 30, 28, 65785))
 
+    def test_timezone_conversion(self):
```
Comment (e-kolpakov) [inline search/tests/tests.py]: @martynjames suggestion: this test doesn't test important edge cases:

* Both have timezones, represent the same moment in time, but timezones are different - should be equal
* Both have timezones, represent different moments in time, local time is equal but timezones are different - should not be equal

File: search/tests/tests.py [hunk 4]
```diff
@@ -551,12 +552,70 @@ def test_pagination(self):
...
+
+        response = self.searcher.search(field_dictionary={"course": "ABC"})
+        self.assertEqual(response["total"], 3)
+        result_ids = [r["data"]["id"] for r in response["results"]]
+        self.assertIn("FAKE_ID_1", result_ids)
+        self.assertIn("FAKE_ID_2", result_ids)
+        self.assertIn("FAKE_ID_3", result_ids)
+        self.assertNotIn("FAKE_ID_11", result_ids)
+        self.assertNotIn("FAKE_ID_12", result_ids)
+        self.assertNotIn("FAKE_ID_13", result_ids)
```
Comment (e-kolpakov) [inline search/tests/tests.py]: @martynjames nit: looks like it's worth writing a helper method

    def assertCollection(self, expected, actual, assertMethod):
        for item in edxpected:
            assertMethod(item, actual)

and use like

    self.assertCollection(["FAKE_ID_1", "FAKE_ID_2", "FAKE_ID_3"], result_ids, self.assertIn)
    self.assertCollection(["FAKE_ID_11", "FAKE_ID_12", "FAKE_ID_13"], result_ids, self.assertNotIn)

File: search/tests/tests.py (hunk 1 above)
Comment (martynjames) [inline search/tests/tests.py]: Yes, but this is before the merge

File: search/tests/mock_search_engine.py (hunk 2 above)
Comment (martynjames) [inline search/tests/mock_search_engine.py]: It was late last night - I will change

File: search/tests/tests.py (hunk 3 above)
Comment (martynjames) [inline search/tests/tests.py]: They are not really important (at least not art the moment). In practice all the times are in UTC; the reason why it was needed is that the json serialisation for file-based operation tips the tzinfo - so sometimes we get tests that include it and sometimes not.

This is only the mock that supports this, so I'd rather not add functionality to the mock where it is not necessary.

File: search/tests/tests.py (hunk 4 above)
Comment (martynjames) [inline search/tests/tests.py]: I disagree - assertIn and assertNotIn are readable items which only requires knowledge of `TestCase` to understand... with the helper method the code gets shorter, but what is actually getting tested is obscured with the helper method.

This is really a difference between test code and product code - I think that test code should be much more verbose, and having things repeat is often a useful consequence to that

**Comment by martynjames** [comment]
Hoping for thumbs up from @dino-cikatic and @e-kolpakov...

File: search/tests/tests.py (hunk 4 above)
Comment (e-kolpakov) [inline search/tests/tests.py]: @martynjames it is a matter of taste anyway, so if you prefer this way - so be it.

**Comment by e-kolpakov** [comment]
@martynjames :+1:

**Comment by dino-cikatic** [comment]
@martynjames LGTM :+1:
